### ISSUE TITLE
No error while saving if validations fails

### ISSUE THREAD

**Issue by dnish** [jagi/meteor-astronomy#613]
Hey,
I'm having an issue when I try to save a document where the validation fails. If I do the following:

    newDoc.save((err) => console.log(err));

the error object is empty, but the document wasn't saved to the database. I just remember that Astronomy also shows the error there if the validations fails. If I do the following, I get the validation error:

    newDoc.validate((err) => console.log(err));

Maybe it has something to do with nested Astronomy classes? In our case we get the following validation error:

```
W20170515-12:27:27.248(2)? (STDERR) { [Error: "metaData.albumId" has to be a string [validation-error]]
W20170515-12:27:27.249(2)? (STDERR)   error: 'validation-error',
W20170515-12:27:27.250(2)? (STDERR)   reason: '"metaData.albumId" has to be a string',
W20170515-12:27:27.250(2)? (STDERR)   details: 
W20170515-12:27:27.250(2)? (STDERR)    [ { className: 'MetaData',
W20170515-12:27:27.251(2)? (STDERR)        type: 'string',
W20170515-12:27:27.251(2)? (STDERR)        name: 'metaData.albumId',
W20170515-12:27:27.251(2)? (STDERR)        nestedName: 'albumId',
W20170515-12:27:27.252(2)? (STDERR)        value: [Object],
W20170515-12:27:27.252(2)? (STDERR)        param: undefined,
W20170515-12:27:27.253(2)? (STDERR)        message: '"metaData.albumId" has to be a string' } ],
W20170515-12:27:27.254(2)? (STDERR)   message: '"metaData.albumId" has to be a string [validation-error]',
W20170515-12:27:27.255(2)? (STDERR)   errorType: 'Meteor.Error' }
```

As you can see, the error comes from "MetaData" which is a sub class in our Astronomy schema.

**Comment by dnish** [comment]
@jagi I've also noticed that when I'm saving an object on server side without callback, and an error occurs, my code below the save method wouldn't be executed. Also there are no error logs in console, so f.e.:

```
const song = new Song({title: 123}); // Validation error, title needs to be String
song.save();
console.log("Document was saved"); // This wouldn't be executed due to the validation error, but we don't get any error information in console. It fails silently. 
```

**Comment by lukejagodzinski** [comment]
Hmmm that would be weird if really happened. Could you create reproduction repository? Without that it's hard to tell.

From validation error I see that you store an array of objects in `metaData.albumId` so validation error is thrown properly. However the `save()` method may not execute any operation if document have not changed. So if you fetched this document from database and try to save it, Astronomy will not do it because it haven't changed from and it's the same as version already stored in collection. So maybe that's a case in your situation.

**Comment by dnish** [comment]
@jagi Yeah I will try to reproduce it. In my second case I get a validation error if I validate the object manually. So f.e:

```
const song = new Song({title: 123});
song.save((err) => { console.log(err);}); // Gives me a validation error
```

But if I do

```
const song = new Song({title: 123});
song.save(); // Execution hangs here
console.log("Saved");
```

the console.log() isn't executed. The execution hangs on the save method, no errors are thrown but the code below it isn't executed. Had the issue several times on different methods.

**Comment by lukejagodzinski** [comment]
So probably there is something wrong in your before or after save, update, insert events. Maybe you're doing some operation that can't be executed and server just does not return any value.

**Comment by dnish** [comment]
Mh okay, I think I'll try to reproduce both cases, starting with one single Astronomy class and then with some nested. Hadn't these problems before, so I agree that it could be caused by some configuration/schema within my project.

**Comment by lukejagodzinski** [comment]
Yep probably, so yes please create reproduction for both cases. It might also be some problem with recent update. But I always run tests so it should not happen.

**Comment by dnish** [comment]
Yep, already tested it with a repro. Couldn't reproduce it, but I saw that I've wrote <code>console.err</code> instead of <code>console.error</code> in my code. After changing it, I see all validations errors and also get the rejection within my promises. Sorry for that 😞

**Comment by lukejagodzinski** [comment]
No problem :)

**Comment by dnish** [comment]
@jagi Run into the same problem again, but this time I found something interesting in my logs. I try to update and remove a object. There is something strange with the "_isNew" property of Astronomy. If I did understand it right, it shows me, if the object was saved to the collection. In my case it is, because I'm querying it from there. But I have a true on it:

```
20170516-16:29:17.372(2)? Class {
I20170516-16:29:17.372(2)?   _isNew: true,
I20170516-16:29:17.372(2)?   _id: 'KBrggg7DKjD6D5Nw7',
...
```

I'm receiving it in my afterRemove event (as e.target). Or is _isNew set to true again after removing it?
